address.md

```diff
@@ -0,0 +1,262 @@
+    Decode(s string) Address
+    Checksum(a Address) []byte
+    Validate(a Address) bool
+}
+```
+##### New()
+
+New returns an Address for the specified network and protocol encapsulating corresponding payload. New fails for unknown network or protocol.
+
+```go
+func New(network byte, protocol byte, payload []byte) Address {
+	if network != Mainnet && network != Testnet {
+        Fatal(ErrUnknownNetwork)
+	}
+	if protocol < SECP256K1 || protocol > BLS {
+        Fatal(ErrUnknownType)
+	}
+    return Address{
+        Network:  network,
+        protocol: protocol,
+        Payload:  payload,
+    }
+}
+```
+
+##### Encode()
+
+Software encoding a Filecoin address must:
+
+- produce an address encoded to a known network
+- produce an address encoded to a known protocol
+- produce an address with a valid checksum
+
+Encodes an Address as a string, converting the network byte to the corresponding prefix and encoding the payload to [base32](https://tools.ietf.org/html/rfc4648).
+
+```go
+func Encode(a Address) string {
+	var prefix string
+    switch a.network {
+    case Mainnet:
+        prefix = Mainnet_Prefix // "f"
+    case Testnet:
+        prefix = Testnet_Prefix // "t"
+    default:
+        Fatal("invalid address network")
+    }
+    
+    switch a.protocol {
+        case SECP256K1:
+        case Actor:
+        case BLS:
+        	cksm := Checksum(a)
+            return prefix + a.protocol + base32.Encode(a.payload + cksm)
+        case ID:
+        	return prefix + a.protocol + base32.Encode(a.payload)
+        default:
+        	Fatal("invalid address protocol")
+    }
+}
+```
+
+##### Decode()
+
+Software decoding a Filecoin address must:
+* verify the network is a known network.
+* verify the protocol is a number of a known protocol.
+
+* verify the checksum is valid
+
+Decode an Address from a string, converting the network prefix and protocol prefix to their corresponding byte values, decoding the payload from base32, and validating the checksum.
+
+```go
+func DecodeString(a string) Address {
+    if len(a) < 4 {
+        Fatal(ErrInvalidLength)
+    }
+    
+    var ntwk byte
+    switch a[:2] {
+    case Mainnet_Prefix:
+        ntwk = Mainnet
+    case Testnet_Prefix:
+    	ntwk = Testnet
+    default:
+        Fatal(ErrUnknownNetwork)
+    }
+    
+    protocol := a[3]
+    payload := a[3:]
+    
+    if protocol == ID {
+        return Address{
+            network: ntwk,
+            protocol: protocol
+            payload: payload
+        }
+    }
+    
+    cksm := payload[len(payload)-CksmLen:len(payload)]
+    
+    if !Validate(a, cksm) {
+        Fatal(ErrInvalidChecksum)
+    }
+    
+    return Address{
+        network: ntwk,
+        protocol: protocol,
+        payload: payload,
+    }
+}
+```
+
+##### Checksum()
+
+Checksum produces a byte array of the blake2b-16 checksum of an address, returning the first 4 bytes of the digest.
+
+```go
+const CksmLen = 4
+
+func Checksum(a Address) [CKSM_LEN]byte {
```

Comment:
The const above is not clearly being used here. Also, are we hashing the entire address struct in this function?